File: Practice/QuickSort.cpp
#include <iostream>
using namespace std;

/**
 * =========================================================
 *                    Quick Sort
 *
 * Time Complexity:
 * Best Case    : O(n log n)
 * Average Case : O(n log n)
 * Worst Case   : O(n²)
 *
 * Space Complexity:
 * Best Case    : O(log n)
 * Average Case : O(log n)
 * Worst Case   : O(n)
 *
 * Stable:
 * No
 *
 * In-Place:
 * Yes
 * =========================================================
 */

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;

    return i + 1;
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pivotIndex = partition(arr, low, high);

        quickSort(arr, low, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, high);
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

int main() {
    int n;

    cout << "Enter the size of the array: ";
    cin >> n;

    int arr[n];

    cout << "Enter " << n << " elements: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    cout << "\nOriginal Array: ";
    printArray(arr, n);

    quickSort(arr, 0, n - 1);

    cout << "Sorted Array: ";
    printArray(arr, n);

    return 0;
}

File: Lecture CodeBase/5EM2/QuickSort.py
def partition(arr, low, high):
    pivot = arr[high]
    i = low -1
    for j in range(low, high):
        if arr[j]<pivot:
            i+=1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1],arr[high] = arr[high],arr[i+1]
    return i+1

def quickSort(arr, low, high):
    if low<high:
        pivotindex = partition(arr, low, high)
        quickSort(arr, low, pivotindex-1)
        quickSort(arr, pivotindex+1, high)

arr = [190, 44, -33, 12, -150, 106, 17, 20]
n = len(arr)
quickSort(arr, 0, n-1)
print(arr)
'''
TC -
    Best Case : O(n log n)
    Average Case : O(n log n)
    Worst Case : O(n)^2

SC - 
    Best Case : O(log n)
    Average Case : O(log n)
    Worst Case : O(n)
'''